pkg/ddevapp/apptypes.go

```diff
@@ -0,0 +1,92 @@
+package ddevapp
+
+type settingsCreator func(*DdevApp) error
+type uploadDir func(*DdevApp) string
+
+// hookSuggestions should probably change its arg from string to app when
+// config refactor is done.
+type hookSuggestions func() []byte
+
+type siteSettingsPaths func(app *DdevApp)
+
+// AppTypeFuncs struct defines the functions that can be called (if populated)
+// for a given appType.
+type AppTypeFuncs struct {
+	settingsCreator
+	uploadDir
+	hookSuggestions
+	siteSettingsPaths
+}
+
+// AppTypeMatrix is a static map that defines the various functions to be called
+// for each apptype (CMS).
+// Example: AppTypeMatrix["drupal"]["7"] == { settingsCreator etc }
+var AppTypeMatrix = map[string]AppTypeFuncs{
+	"php": AppTypeFuncs{},
+	"drupal7": AppTypeFuncs{
+		CreateDrupalSettingsFile, GetDrupalUploadDir, GetDrupal7Hooks, SetDrupalSiteSettingsPaths,
+	},
+	"drupal8": AppTypeFuncs{
+		CreateDrupalSettingsFile, GetDrupalUploadDir, GetDrupal8Hooks, SetDrupalSiteSettingsPaths,
+	},
+	"wordpress": AppTypeFuncs{
+		CreateWordpressSettingsFile, nil, GetWordpressHooks, SetWordpressSiteSettingsPaths,
+	},
+	"backdrop": AppTypeFuncs{},
+	"typo3":    AppTypeFuncs{},
+}
+
+// GetValidAppTypes returns the valid apptype keys from the AppTypeMatrix
+func GetValidAppTypes() []string {
+	keys := make([]string, 0, len(AppTypeMatrix))
+	for k := range AppTypeMatrix {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// IsValidAppType checks to see if the given apptype string is a valid configured
+// apptype.
+func IsValidAppType(apptype string) bool {
+	if _, ok := AppTypeMatrix[apptype]; ok {
+		return true
+	}
+	return false
+}
+
+// CreateSettingsFile creates the settings file (like settings.php) for the
+// provided app is the apptype has a settingsCreator function.
+func CreateSettingsFile(l *DdevApp) error {
+	if appFuncs, ok := AppTypeMatrix[l.GetType()]; ok && appFuncs.settingsCreator != nil {
+		err := appFuncs.settingsCreator(l)
+		return err
+	}
+	return nil
+}
+
+// UploadDirFunc returns the upload (public files) directory for the given app
+func UploadDirFunc(l *DdevApp) string {
+	if appFuncs, ok := AppTypeMatrix[l.GetType()]; ok && appFuncs.uploadDir != nil {
+		uploadDir := appFuncs.uploadDir(l)
+		return uploadDir
+	}
+	return ""
+}
+
+// GetHookSuggestions gets the actual text of the config.yaml hook suggestions
+// for a given apptype
+func GetHookSuggestions(apptype string) []byte {
```

Comment:
I'm not super clear on what this returns. Is this something that's going to be different for each app type?

Reply:
Ah, I see, looking down at the implementation, this is for suggested hook settings (so `drush cr` after a db import or whatever). Given that "hook suggestions" is a phrase that has some pretty specific meaning to Drupal developers, can we rename this to something else? Maybe `GetSuggestedHooks` or `GetDefaultHooks` or something like that?

Reply:
Makes great sense.